packages/babel-plugin-transform-react-display-name/src/index.js

```diff
@@ -42,53 +69,161 @@ export default function ({ types: t }) {
+                     || path.parentPath.node.type == "ExportNamedDeclaration") {
+            setDisplayNameAfter(path, path.node.id, displayName);
+          }
+        }
+      },
+      FunctionExpression: function (path) {
+        if (shouldSetDisplayNameForFuncExpr(path)) {
+          const id = findCandidateNameForExpression(path);
+          if (id) {
+            setDisplayNameAfter(path, id);
+          }
+        }
+      },
+      ArrowFunctionExpression: function (path) {
+        if (shouldSetDisplayNameForFuncExpr(path)) {
+          const id = findCandidateNameForExpression(path);
+          if (id) {
+            setDisplayNameAfter(path, id);
+          }
         }
       },
     },
   };
 }
+
+function shouldSetDisplayNameForFuncExpr(path) {
+  // Parent must be either 'AssignmentExpression' or
+  // 'VariableDeclarator' or 'CallExpression' with a parent of 'VariableDeclarator'
+  if (path.parentPath.node.type === "AssignmentExpression" &&
+      path.parentPath.node.left.type !== "MemberExpression" && // skip static members
+      path.parentPath.parentPath.node.type == "ExpressionStatement" &&
+      path.parentPath.parentPath.parentPath.node.type == "Program") {
+    return doesReturnJSX(path.node.body);
+  } else {
+    // if parent is a call expression, we have something like (function () { .. })()
+    // move up, past the call expression and run the rest of the checks as usual
+    if (path.parentPath.node.type === "CallExpression") {
+      path = path.parentPath;
+    }
+
+    if (path.parentPath.node.type === "VariableDeclarator") {
+      if (path.parentPath.parentPath.parentPath.node.type === "ExportNamedDeclaration" ||
+          path.parentPath.parentPath.parentPath.node.type === "Program") {
+        return doesReturnJSX(path.node.body);
+      }
+    }
+  }
+
+  return false;
+}
+
+function classHasRenderMethod(node) {
+  if (!node.node.body) {
+    return false;
+  }
+  const members = node.node.body.body;
+  for (let i = 0; i < members.length; i++) {
+    if (members[i].type == "ClassMethod" && members[i].key.name == "render") {
+      return true;
+    }
+  }
+
+  return false;
+}
+
+// crawl up the ancestry looking for possible candidates for displayName inference
+// optionally look for object properties
+function findCandidateNameForExpression(path, inclObjProp) {
+  let id;
+  path.find(function (path) {
+    if (path.isAssignmentExpression()) {
+      id = path.node.left;
+    } else if (inclObjProp && path.isObjectProperty()) {
+      id = path.node.key;
+    } else if (path.isVariableDeclarator()) {
+      id = path.node.id;
+    } else if (path.isStatement()) {
+      // we've hit a statement, we should stop crawling up
+      return true;
+    }
+
+    // we've got an id! no need to continue
+    if (id) return true;
+  });
+  return id;
+}
+
+function doesReturnJSX (body) {
+  if (!body) return false;
+  if (body.type === "JSXElement") {
```

Comment:
`null` is also a valid JSX type, and with React 16 so will an array of `JSXElement`; though missing these for now is probably fine as it's just a heuristic.